core/serialization/blocks.js

```diff
@@ -0,0 +1,143 @@
+ *       serialized. False by default.
+ *     addCoordinates: If true the coordinates of the block are added to the
+ *       serialized state. False by default.
+ * @return {?Blockly.serialization.blocks.State} The serialized state of the
+ *     block, or null if the block could not be serialied (eg it was an
+ *     insertion marker).
+ */
+const save = function(block, {addCoordinates = false} = {}) {
+  if (block.isInsertionMarker()) {
+    return null;
+  }
+
+  const state = Object.create(null);
+  state['type'] = block.type;
+  state['id'] = block.id;
+
+  if (addCoordinates) {
+    addCoords(block, state);
+  }
+  addAttributes(block, state);
+  addExtraState(block, state);
+
+  return state;
+};
+
+/**
+ * Adds attributes to the given state object based on the state of the block.
+ * Eg collapsed, disabled, editable, etc.
+ * @param {!Blockly.Block} block The block to base the attributes on.
+ * @param {!Blockly.serialization.blocks.State} state The state object to append
+ *     to.
+ */
+const addAttributes = function(block, state) {
+  if (block.isCollapsed()) {
+    state['collapsed'] = true;
+  }
+  if (!block.isEnabled()) {
+    state['disabled'] = true;
+  }
+  if (!block.isEditable()) {
+    state['editable'] = false;
+  }
+  if (!block.isDeletable() && !block.isShadow()) {
+    state['deletable'] = false;
+  }
+  if (!block.isMovable() && !block.isShadow()) {
+    state['movable'] = false;
+  }
+
+  if (block.inputsInline !== undefined &&
+      block.inputsInline !== block.inputsInlineDefault) {
+    state['inline'] = block.inputsInline;
+  }
+
+  // Data is a nullable string, so we don't need to worry about falsy values.
+  if (block.data) {
+    state['data'] = block.data;
+  }
+};
+
+/**
+ * Adds the coordinates of the given block to the given state object.
+ * @param {!Blockly.Block} block The block to base the coordinates on.
+ * @param {!Blockly.serialization.blocks.State} state The state object to append
+ *     to.
+ */
+const addCoords = function(block, state) {
+  const workspace = block.workspace;
+  const xy = block.getRelativeToSurfaceXY();
+  state['x'] = Math.round(workspace.RTL ? workspace.getWidth() - xy.x : xy.x);
+  state['y'] = Math.round(xy.y);
+};
+
+/**
+ * Adds any extra state the block may provide to the given state object.
+ * @param {!Blockly.Block} block The block to serialize the extra state of.
+ * @param {!Blockly.serialization.blocks.State} state The state object to append
+ *     to.
+ */
+const addExtraState = function(block, state) {
+  if (block.saveExtraState) {
+    const extraState = block.saveExtraState();
+    if (extraState !== null) {
+      state['extraState'] = extraState;
+    }
+  } else if (block.mutationToDom) {  // Backwards compatibility.
+    const mutation = block.mutationToDom();
+    if (mutation !== null) {
+      state['extraState'] = Blockly.Xml.domToText(mutation)
+          .replace('xmlns="https://developers.google.com/blockly/xml"', '');
+ 
+    }
```

Comment:
Instead of handle the backwawrds compatibility here, we could provide a script w/ the release that automatically add wrapper implementations to people's block definitions.

This might be better because it would simplify this logic, and make it so the JSO system doesn't rely on the XML system. (Plus I might get to break out my python skillz :D )

Reply:
Discussed this on the other PR and we decided to remove backwards compatibility for now.